### ISSUE TITLE
Unable to obtain the final result.

### ISSUE THREAD

**Issue by wff-R** [rwanwork/sam2irs#4]
Hello,
I have a small suggestion. When configuring this environment, conda informed me that the Intel channel is taken off the shelf. So I used https://software.repos.intel.com/python/conda instead and successfully installed it. Finally, the example can run normally. I think the parameters in environment.yml can be changed to: https://software.repos.intel.com/python/conda.
However, when I run my own data, the final gtf file is always blank. My code is as follows: samtools view /home/data/wf/RNAseq/LTR/BAM/PT3-CON-1.markdup.sorted.bam | /home/data/software/sam2irs/sam2irs/Perl/sam2irs.pl --verbose 1 --chrlist /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes --gtf /home/data/wf/RNAseq/LTR/sam2irs/gencode.vM25_protein_coding_exon_sam2irs.gtf 1>/home/data/wf/RNAseq/LTR/sam2irs/test.gtf 2>/home/data/wf/RNAseq/LTR/sam2irs/test.error.txt.
In addition, I attach the file I used to run. 
BAM file:
https://drive.google.com/file/d/1BE_fFIdE2jvx6LmgTTV7p-YRP51BrtZ9/view?usp=sharing
gtf file:
https://drive.google.com/file/d/14EYeJTBYTBvkrj4tMmUK9NX6RYlKG6qx/view?usp=drive_link
Chromosome list:
https://drive.google.com/file/d/1IdMXIsxiRE5YKbRVLZr7cHSGja5Uhzx4/view?usp=drive_link
test.error:
https://drive.google.com/file/d/1XMsKkbHku-mKtEiXtfwZQYPx-O3m_ttD/view?usp=drive_link

I have tried for a long time but still haven't been able to solve it. I hope you can help me see where the problem lies.

**Comment by wff-R** [comment]
hello,
Sorry, I have found my problem. It is because the format of the GTF I input is incorrect. After changing it, it is running normally. Thank you very much for sharing the code.

**Comment by rwanwork** [comment]
> hello, Sorry, I have found my problem. It is because the format of the GTF I input is incorrect. After changing it, it is running normally. Thank you very much for sharing the code.

Thank you for your interest in our work!  Indeed, I only recently (1-2 weeks ago) noticed the problem with the Intel channel.  I will fix that.

Also, I'm glad you were able to find the problem.  Even so, I will keep this ticket open; I'd like to see the problem myself and see if I can update the code to produce a better, more informative error message based on your GTF file.  If you don't mind, can you keep your test files in Google Drive?  I'll try to take a look some time this week and get back to you.

**Comment by rwanwork** [comment]
I just took a look at the 4 test files you uploaded and it seems they have been removed.  If you can put them back and/or explain what you had wrong with the GTF file, I can update the script to try to deal with it (depends on what was the issue).

Otherwise, I think I will close the issue as there is nothing else I can do.

As for `environment.yml`, the `intel` channel is actually unnecessary.  I should have removed it a long time ago.  I've just removed it with the latest commit.  Should be ok now.

**Comment by wff-R** [comment]
I hope this email finds you well. First of all, I would like to apologize for not replying in time yesterday as it was the Mid-Autumn Festival.

I am pleased to inform you that I have just tried installing the software, and it went extremely smoothly. There are currently no issues. As I am just starting out in bioinformatics and this was my first time uploading the raw data to Google Drive. I did not delete the data on my end. I think there might have been some unknown issues during my first attempt due to my inexperience. The previous error was entirely due to my misunderstanding that the file I exported was in gtf format, but it actually wasn't. After inputting the correct gtf format file, the software is running very smoothly. So, the previous problem did not involve the need to update the code.

I am very grateful to your team for generously sharing the code and being able to provide such prompt maintenance.

Furthermore, I have a small question. When I run my own data, the log file outputs some statements like "Use of uninitialized value&nbsp;gtf_fp&gt; line 1 (#1)". I am not sure what causes this. Will this have an impact on the final result? The log file is attached to this email. Thank you very much.



Fan Wang


王凡
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "rwanwork/sam2irs"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年9月18日(星期三) 凌晨0:40
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [rwanwork/sam2irs] Unable to obtain the final result. (Issue #4)





 
I just took a look at the 4 test files you uploaded and it seems they have been removed.  If you can put them back and/or explain what you had wrong with the GTF file, I can update the script to try to deal with it (depends on what was the issue).
 
Otherwise, I think I will close the issue as there is nothing else I can do.
 
As for environment.yml, the intel channel is actually unnecessary.  I should have removed it a long time ago.  I've just removed it with the latest commit.  Should be ok now.

**Comment by rwanwork** [comment]
No worries!

I think the permission on the files you shared on Google Drive were not correct, so I was unable to download them.  But since you said that problem is fixed, we can leave it alone.

As for your second question about the "Use of uninitialized value gtf_fp>; line 1", that is potentially a problem.  It could be a problem with my program.  Can you give me some details; perhaps the input files that you are using?  If the files are big, you can reduced them to the smallest size necessary to reproduce the problem.  For example, `head -n 10 <file>` would produce a new file that is only the first 10 lines of the original.

I think only after seeing where the problem is from will I be answer your question about whether or not the problem is serious.  I've just re-opened this Issue.

**Comment by wff-R** [comment]
I apologize for being new to Google Drive. I’ve tried resetting the permissions, and it should be downloadable now. I followed the software instructions, and the software ran very smoothly. However, the error in the run log is the same as I mentioned before. When calculating the IR score based on the final GTF file, I found it a bit strange that for one sample, all the aligned reads and aligned bases for introns are exactly the same. Since the aligned bases are very large, it results in an extremely small IR score. I have uploaded the corresponding result files, and I would appreciate it if you could take a look and let me know if there’s any issue.
BAM file:
https://drive.google.com/file/d/1BE_fFIdE2jvx6LmgTTV7p-YRP51BrtZ9/view?usp=drive_link
gtf file:
https://drive.google.com/file/d/14EYeJTBYTBvkrj4tMmUK9NX6RYlKG6qx/view?usp=drive_link
Chromosome list:
https://drive.google.com/file/d/1IdMXIsxiRE5YKbRVLZr7cHSGja5Uhzx4/view?usp=drive_link
test.error:
https://drive.google.com/file/d/1BhmnLn6Gb_bRNiKpSJuu13gTrb6ilUQZ/view?usp=drive_link
gtf result file:
https://drive.google.com/file/d/1XCkxnXT0ANvyZwBNKJDHcDTjX07La3qr/view?usp=drive_link


Thank you!

**Comment by rwanwork** [comment]
Thanks!  I've requested access to these files.  Can you give them to me?  You can confirm my e-mail address at [here](https://github.com/rwanwork/), of course (on the left).

**Comment by wff-R** [comment]
It has been shared to another email address. Thank you.

**Comment by rwanwork** [comment]
Thank you!  I have downloaded all of the files.  You can delete them.

I don't have the same problem as you have, but I am encountering a problem.  Let me see if it is a problem with my program and, if so, try to fix it.  I'll try to get back to you later this week; I'm busy with something else at the moment.

**Comment by rwanwork** [comment]
Thanks for providing the test files.  I did find some issues with the program and have corrected it.

Can you get the latest version and try again?  Hopefully, you won't get any of the errors or warnings that you had before.  If you do, please let me know.

**Comment by wff-R** [comment]
Hello! I have reinstalled the latest software and tried to run it once. According to the log file, the alignments of chromosomes not included in the gtf file should have been deleted. The other running results are the same as before. The log file is as follows：
II      Filenames
II        Input chromosome list:  /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes
II        Input GTF file:  /home/data/wf/RNAseq/LTR/sam2irs/gencode.vM25.primary_assembly.annotation.gtf
II        Record of SAM file:  Not provided
II        Log file:  Not provided
II      Output exons:  No

Use of uninitialized value $attribute_tmp in pattern match (m//) at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 324, <$gtf_fp> line 1 (#1)
    (W uninitialized) An undefined value was used as if it were already
    defined.  It was interpreted as a "" or a 0, but maybe it was a mistake.
    To suppress this warning assign a defined value to your variables.
    
    To help you figure out what was undefined, perl will try to tell you
    the name of the variable (if any) that was undefined.  In some cases
    it cannot do this, so it also tells you what operation you used the
    undefined value in.  Note, however, that perl optimizes your program
    and the operation displayed in the warning may not necessarily appear
    literally in your program.  For example, "that $foo" is usually
    optimized into "that " . $foo, and the warning will refer to the
    concatenation (.) operator, even though there is no . in
    your program.
    
Use of uninitialized value $feature_tmp in string eq at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 330, <$gtf_fp> line 1 (#1)
Use of uninitialized value $feature_tmp in string eq at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 337, <$gtf_fp> line 1 (#1)
Use of uninitialized value $feature_tmp in string eq at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 340, <$gtf_fp> line 1 (#1)
Use of uninitialized value $feature_tmp in string eq at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 343, <$gtf_fp> line 1 (#1)
Use of uninitialized value $attribute_tmp in pattern match (m//) at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 324, <$gtf_fp> line 2 (#1)
Use of uninitialized value $feature_tmp in string eq at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 330, <$gtf_fp> line 2 (#1)
Use of uninitialized value $feature_tmp in string eq at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 337, <$gtf_fp> line 2 (#1)
Use of uninitialized value $feature_tmp in string eq at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 340, <$gtf_fp> line 2 (#1)
Use of uninitialized value $feature_tmp in string eq at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 343, <$gtf_fp> line 2 (#1)
Use of uninitialized value $attribute_tmp in pattern match (m//) at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 324, <$gtf_fp> line 3 (#1)
Use of uninitialized value $feature_tmp in string eq at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 330, <$gtf_fp> line 3 (#1)
Use of uninitialized value $feature_tmp in string eq at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 337, <$gtf_fp> line 3 (#1)
Use of uninitialized value $feature_tmp in string eq at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 340, <$gtf_fp> line 3 (#1)
Use of uninitialized value $feature_tmp in string eq at
        /home/data/software/sam2irs/Perl/sam2irs.pl line 343, <$gtf_fp> line 3 (#1)
II      Processed GTF file (/home/data/wf/RNAseq/LTR/sam2irs/gencode.vM25.primary_assembly.annotation.gtf, 843715 records):
II        Retained ('exon'):            843712
II        Discarded:
II          Gene names with 'Gm':       0
II          # of 'CDS':                 0
II          # of 'start_codon':         0
II          # of 'stop_codon':          0
II          # of Others:                3

II      SAM file (STDIN, 45788347 records):
II        Number of header lines:               0
II        Number of incomplete records:         0
II        Number of records stored:             45788347
II      Processing chromosome lengths (/home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes)...

II      Total chromosomes to process:  49
II      Chromosome GL456210.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456211.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456212.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456213.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456216.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456219.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456221.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456233.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456350.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456354.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456359.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456366.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456367.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456370.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456372.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456378.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456379.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome GL456382.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome JH584293.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome JH584294.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome JH584295.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome JH584296.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome JH584297.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome JH584298.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome JH584299.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome JH584301.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Chromosome JH584304.1 not in /home/data/wf/RNAseq/LTR/sam2irs/mm10.chrom.sizes ... skipping.
II      Processing chr1 ...
II      Processing chr10 ...
II      Processing chr11 ...
II      Processing chr12 ...
II      Processing chr13 ...
II      Processing chr14 ...
II      Processing chr15 ...
II      Processing chr16 ...
II      Processing chr17 ...
II      Processing chr18 ...
II      10 of 49 chromosomes processed...
II      Processing chr19 ...
II      Processing chr2 ...
II      Processing chr3 ...
II      Processing chr4 ...
II      Processing chr5 ...
II      Processing chr6 ...
II      Processing chr7 ...
II      Processing chr8 ...
II      Processing chr9 ...
II      Processing chrM ...
II      20 of 49 chromosomes processed...
II      Processing chrX ...
II      Processing chrY ...

II      Total SAM records processed:  45788347
II        Intron records:  346967
II        Exon records:  20320158
II        Intron+Exon records:  156545
II        Unaligned records:  0
II        Unknown records:  24964677

The log file still uses "Use of uninitialized value". I don't know if this will affect the final result. In addition, the result I obtained still makes me a little confused. The aligned_bases of all introns are the same value. My understanding is that it is for the number of bases that are matched on the input chromosome. According to the calculation formula, The intron retention score is: (15 / 7) / 36 = 0.0595. Then the IR score should be very small because the number of bases aligned to the intron is much smaller than the total aligned_bases. I'll give an example. I randomly pick out some intron retention regions from the final document. For this intron retention region, its IR score should be 296/7112/6750803178，a very small value, and this value will definitely not be greater than 1. However, the distribution range of log2(IR score) in the article is between -10 and 10. I always feel that the IR score in the article will not be as small as in my data. But I don't know where the problem lies. Thank you for your patient answers.

**Comment by rwanwork** [comment]
Thanks for the prompt feedback!

`Use of uninitialized value` is most likely a problem with the program.  It is either a bug or there is something in the input that my program was not ready for.  So, I definitely should fix it.  For example, your GTF file should have a single tab character between each field; if you had spaces instead, it would generate this kind of error.

I didn't see the above error when I ran the program with your sample data.  Are you using the exact same inputs or has something changed?  If so, let me try again tonight.  If not, perhaps you can send to me the first 10 lines of it (i.e., `head -n 10 gencode.vM25.primary_assembly.annotation.gtf >newfile.gtf`)?

As for your second concern, yes, you are correct.  The IR score itself does not matter.  What matters is the relative score between one intron and another intron within your sample.  So, for our paper, we did normalise the score in order to get "sensible" values.  The score is calculated using `score-irs.pl`.  You can provide different values for the normalising constant by using the `--normaliser` option.  Try different ones until you get the range of values that looks right for you.

If you have 2 samples (for example) such as a case and control, make sure for both samples, you use the same value for `--normaliser`.  In our paper, we had various samples like quiescent, activated, etc.  All of them used the normaliser `100000000000`.  There is nothing scientific about this value; it was just something to give all values within a particular range.

I think this is what you're asking me?  Anyway, let me fix the program first and get back to you.  If your input GTF is not the same as what you sent to me earlier, please send me the first ~10 lines of it to help me investigate.  Thank you!

**Comment by wff-R** [comment]
Hello! I am extremely sorry. After verification, I found that the GTF file I input is indeed not the same as the one uploaded before, and the format has changed slightly. After I tried to replace it with a new GTF file, there was no abnormal output in the operation log.
I am very grateful for your detailed answer to my second question. I fully understand and know what to do next. Thank you very much for your help. Since I didn't see relevant introductions in the Readme, the subsequent calculations were made by myself according to the introduction in the Readme. The code you shared is very useful to me. However, I am not very familiar with bioinformatics. In fact, I was curious about the functions of the other scripts under the Perl folder before, but I didn't find relevant introductions, so I didn't learn about them. After communicating, I think your program is very complete. It would be even more perfect if you could add introductions to the other several scripts. Thank you again for your help.

**Comment by rwanwork** [comment]
No problem!

If you have some input file that produces that warning / error again, do not hesitate to let me know.  Your feedback will let me see what kind of input files are out there and help me make sure my scripts are more robust.

As for your second point, you are indeed correct.  The focus of this repository was `sam2irs.pl`.  The other scripts were added in to make that script's output more "useful".  But you are correct -- they are important, too, and there is currently too little documentation about them.  I will add some more information this or next week (when I am free) and update the repository.  Feel free to check back to see the update.

I have created Issue #5 for myself as a reminder to do this.  When this is done, I will close that issue.

Thanks again for your interest in our work and your patience in helping me test it!
